data/ctpl.1
.TH CTPL 1 "15 Apr 2010"
.SH NAME
\fBctpl\fP \- CTPL command-line template parsing tool

.SH SYNOPSIS
.B ctpl
[\fIOPTION\fR]... \fIINPUTFILE\fR...
.SH DESCRIPTION
.B ctpl
parses one or more \fIINPUTFILE\fR template against the provided environment and
outputs the result.

.LP
.SH OPTIONS

.TP
\fB\-h\fR, \fB\-\-help\fR
Prints the command line options.

.TP
\fB\-o\fR, \fB\-\-output\fR=\fIFILE\fR
Write output to \fIFILE\fR. If not provided, defaults to stdout.

.TP
\fB\-e\fR, \fB\-\-env\-file\fR=\fIENVFILE\fR
Add environment from \fIENVFILE\fR. This option may appear more than once.

.TP
\fB\-c\fR, \fB\-\-env\-chunk\fR=\fICHUNK\fR
Add environment chunk \fICHUNK\fR. This option may appear more than once.

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Be verbose.

.TP
\fB\-\-version\fR
Print the version information and exit.

.TP
\fB\-\-encoding\fR=\fIENCODING\fR
Specify the encoding of the input and output files. The default encoding is the
system's one.

.SH TEMPLATE AND ENVIRONMENT DESCRIPTION SYNTAX
For the documentation about the syntax of templates and environment
descriptions, see the CTPL library's documentation.

.SH REPORTING BUGS
Report CTPL bugs to ban@herbesfolles.org
.br
CTPL home page: <http://ctpl.tuxfamily.org>

.SH AUTHOR
Colomban Wendling (ban@herbesfolles.org)

.SH COPYRIGHT
Copyright \(co 2010 Colomban Wendling.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# 0.3

  The version 0.3 is an API and ABI break from the previous 0.2 branch.
  
  ## Changes summary
  
  * Heavy API cleanup;
  * Array expressions can now be indexed;
  * The `for` loop now iterates over an expression rather than a symbol name;
  * CtplEnviron now uses reference counting;
  * Improvements of CtplEnviron's memory usage;
  * Bug fixes:
    * ctpl_value_copy(), ctpl_value_set_string() and ctpl_value_set_array*() can
      now take a part of itself as the value to set/copy;
    * Fix internal handling of an error which may lead to wrong error detection;
    * Fix wrong name of streams built from GFiles, paths and URIs;
    * Fix dramatic slowness when a symbol gets pushed many times;
  * Some internal refactoring.
  
  Yes, it is the second API break since the library's birth, less than one year
  ago. I know it and I apologize for the inconvenience.
  
  But even if it's an API break, if you make normal usage of the library, it
  shouldn't be too painful for you to upgrade to the new version (see below).
  Many functions and data structures have been removed or made private, but the
  API that was meant to be used by final users had only small changes.
  
  I will also justify this new break, at least for you to understand the
  original motivations for it:
  * There was way too many symbols exposed, and a lot of them wasn't supposed to
    be used by library's users;
  * Exposing data structures (as well as some other internals) makes working on
    the core harder, because any change to one of this data structure means an
    API break.
  Then this new break is here to -- hopefully -- ease future improvements of
  CTPL, as well as preventing more API breaks.
  
  ## What to change in your code
  
  ### Token-related functions
  
  The functions ctpl_token_free() and ctpl_token_expr_free() lost their second
  (boolean) argument. If you passed TRUE here, simply remove that second
  argument. Otherwise, you did strange things and you're screwed (but see the
  next section).
  
  ctpl_lexer_free_tree() has also been removed, use ctpl_token_free() instead
  (the semantic is exactly the same).
  
  ### Environments
  
  CtplEnviron now uses reference counting. It means that you get all the good
  of reference counting, but you also have to replace calls to
  ctpl_environ_free() with ctpl_environ_unref(). Note that ctpl_environ_unref()
  does not support NULL argument.
  
  ctpl_environ_pop() has changed. It now takes a third argument to get the
  poped value and returns a boolean (see the documentation for more details).
  Anyway you should probably not need ctpl_environ_pop() at all.
  
  ### Input and output streams
  
  ctpl_input_stream_read_double() and ctpl_input_stream_read_long() are renamed
  to ctpl_input_stream_read_float() and ctpl_input_stream_read_int() for more
  consistency with the rest of the API.
  
  ### Others
  
  The environment to pass to ctpl_eval_*() is not constant anymore. This will
  probably not change anything for you, but it's a subtle change. This is to
  allow future operators to affect the environment if needed (which isn't the
  case yet).
  
  ## The dark side of the cleanup
  
  Apart the inconvenience of an API break, some "features" of the library may be
  gone -- yes, having many internal exposed may allow unwanted "features".
  First, know that these features weren't planned. Then, feel free to contact
  me, explaining what functionality you miss and why, and I'll probably be happy
  to re-introduce it cleanly for you.
  
  ## CLI tool changes
  
  The CLI tool, apart then new features of the template engine, gained support
  for encodings. This brings a new `--encoding` option allowing to specify the
  encoding of the input and output files. The default encoding is the system's
  one.


# 0.2.2 (25/05/2010)

  The version 0.2.2 is a bugfix release for the 0.2 branch.
  
  ## Changes summary
  
  * Fix lexing of a whole expression input (using ctpl_lexer_expr_lex() or
    one of its variants);
  * Fix priority of boolean operators AND and OR;
  * Fix support of operators <= and >=;
  * Better error reporting in the expression lexer;
  * New alternative build system: Waf (from a work by Enrico Tröger, thanks).


# 0.2.1 (27/04/2010)

  The version 0.2.1 is a bugfix release for the 0.2 branch.
  
  ## Changes summary
  
  * Do not export the private API;
  * Little cleaning of the build system.


# 0.2 (15/04/2010)
  
  The version 0.2 is an API and ABI break from the previous 0.1 branch.
  
  ## Changes summary (in-depth):
  
  ### New features and important changes
  
  * GIO replaces LibMB. Impact is:
    * Less memory usage (input template may or may not be loaded into memory);
    * Small performance regression, between 8% to 19% [1] (on extreme cases),
      measured on heavy templates;
    * Any GIO stream may be live-lexed, from local files to remote FTP, Samba,
      DAV, etc.
  * Better lexer error reporting, with line and position marks;
  * Improved numeric constants loading, including:
    * Integers are now loaded as integers, impacting:
      * No more precision loss;
      * Shorter maximum value, but still more than reasonable (C's long: in 64
        bits machines should be something like from -9223372036854775807 to
        +9223372036854775807, and -2147483647 to +2147483647 on 32 bits
        machines).
    * Support for octal and binary constants via 0o and 0b prefix respectively;
  * Improved token representation, saving about from 19% to 28% of memory use
    for the token tree (no measurable performance impact) (already in 0.1.90);
  * Add support for single-line comments in environ descriptions;
  * Fix support of array values that starts with blanks in environ descriptions;
  * `make distcheck` and out-of-tree build now work;
  * Add man page for the ctpl command-line utility.
  
  As you can see, the main improvements of this release are less memory

src/ctpl.c
/* 
 * 
 * Copyright (C) 2009-2010 Colomban Wendling <ban@herbesfolles.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <locale.h>
#include <unistd.h> /* for STDOUT_FILENO */
#include <glib.h>
#include <gio/gio.h>
#include <gio/gunixoutputstream.h>
#include "ctpl.h"


#define GETTEXT_PACKAGE NULL /* FIXME: */


/* options */
static gchar      **OPT_env_files     = NULL;
static gchar      **OPT_env_chunks    = NULL;
static gchar      **OPT_input_files   = NULL;
static gchar       *OPT_output_file   = NULL;
static gboolean     OPT_verbose       = FALSE;
static gboolean     OPT_print_version = FALSE;
static gchar       *OPT_encoding      = NULL;

static GOptionEntry option_entries[] = {
  { "output", 'o', 0, G_OPTION_ARG_FILENAME, &OPT_output_file,
    "Write output to FILE. If not provided, defaults to stdout.", "FILE" },
  { "env-file", 'e', 0, G_OPTION_ARG_FILENAME_ARRAY, &OPT_env_files,
    "Add environment from ENVFILE. This option may appear more than once.",
    "ENVFILE" },
  { "env-chunk", 'c', 0, G_OPTION_ARG_STRING_ARRAY, &OPT_env_chunks,
    "Add environment chunk CHUNK. This option may appear more than once.",
    "CHUNK" },
  { "verbose", 'v', 0, G_OPTION_ARG_NONE, &OPT_verbose,
    "Be verbose.", NULL },
  { "version", 0, 0, G_OPTION_ARG_NONE, &OPT_print_version,
    "Print the version information and exit.", NULL },
  { "encoding", 0, 0, G_OPTION_ARG_STRING, &OPT_encoding,
    "Specify the encoding of the input and output files.", "ENCODING" },
  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &OPT_input_files,
    "Input files", "INPUTFILE[...]" },
  { NULL }
};


/* prints verbose messages */
static void G_GNUC_PRINTF(1, 2)
printv (const gchar *fmt,
        ...)
{
  if (OPT_verbose) {
    va_list ap;
    
    va_start (ap, fmt);
    vfprintf (stdout, fmt, ap);
    va_end (ap);
  }
}

/* prints error messages */
static void G_GNUC_PRINTF(1, 2)
printerr (const gchar *fmt,
          ...)
{
  va_list ap;
  
  va_start (ap, fmt);
  vfprintf (stderr, fmt, ap);
  va_end (ap);
}

/* checks whether an encoding is compatible with the ASCII charset, so whether
 * a conversion is needed. return %TRUE if conversion is needed, %FALSE
 * otherwise */
static gboolean
encoding_needs_conversion (const gchar *encoding)
{
  static GRegex *re = NULL;
  
  if (! re) {
    GError *err = NULL;
    
    /* TODO: check current encodings and add more */
    re = g_regex_new ("^("
                      "(US-|cs)?ASCII([-_ ]?[78])"                           "|"
                      "US"                                                   "|"
                      "ANSI[-_ ]X3.4[-_ ]19(68|86)"                          "|"
                      "ISO[-_ ]?646([-_ ]?US|\\.irv[-_ :]?1991)"             "|"
                      "ISO[-_ ]IR[-_ ]6"                                     "|"
                      "(IBM|CP|OEM)[-_ ]?(367|437|737|850|858|869)"          "|"
                      "utf[-_ ]?8"                                           "|"
                      "iso(/cei)?[-_ ]?8859.*"                               "|"
                      "(windows|CP)[-_ ]1252"
                      ")$",
                      G_REGEX_CASELESS | G_REGEX_NO_AUTO_CAPTURE, 0, &err);
    /* should never happen, means that the code is wrong */
    if (err) {
      g_critical ("Internal error: failed to build regular expression for "
                  "matching encoding: %s", err->message);
      g_error_free (err);
    }
  }
  
  return ! g_regex_match (re, encoding, 0, NULL);
}

/* parses the options and fills OPT_* */
static gboolean
parse_options (gint    *argc,
               gchar ***argv,
               GError **error)
{
  gboolean        success = FALSE;
  GOptionContext *context;
  
  context = g_option_context_new ("- CTPL template parser");
  g_option_context_add_main_entries (context, option_entries, GETTEXT_PACKAGE);
  if (g_option_context_parse (context, argc, argv, error)) {
    if (OPT_print_version) {
      printf ("CTPL %s\n", VERSION);
      exit (0);
    } else if (OPT_input_files == NULL) {
      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
                   "Missing input file(s)");
    } else {
      if (! OPT_encoding) {
        const gchar *local_charset;
        
        /* if no encoding was given, assume it is the syetem one, so setup the
         * encoding to convert if needed */
        g_get_charset (&local_charset);
        OPT_encoding = g_strdup (local_charset);
      }
      success = TRUE;
    }
  }
  g_option_context_free (context);
  
  return success;
}

/* Creates a CtplInputStream from a command-line argument */
static CtplInputStream *
open_input_stream (const gchar *arg,
                   GError     **error)
{
  GFile            *file;
  CtplInputStream  *stream = NULL;
  GInputStream     *gstream = NULL;
  GFileInputStream *gfstream;
  
  file = g_file_new_for_commandline_arg (arg);
  gfstream = g_file_read (file, NULL, error);
  if (gfstream) {
    gstream = G_INPUT_STREAM (gfstream);
    
    if (encoding_needs_conversion (OPT_encoding)) {
      GCharsetConverter *converter;
      
      converter = g_charset_converter_new ("utf8", OPT_encoding, error);
      if (! converter) {
        g_object_unref (gstream);
        gstream = NULL;
      } else {
        GInputStream *gcstream;
        
        gcstream = g_converter_input_stream_new (G_INPUT_STREAM (gstream),
                                                 G_CONVERTER (converter));
        g_object_unref (gstream);
        gstream = gcstream;
        g_object_unref (converter);
      }
    }
  }
  g_object_unref (file);
  if (gstream) {
    gchar *name = g_filename_display_basename (arg);
    
    stream = ctpl_input_stream_new (gstream, name);
    g_free (name);
    g_object_unref (gstream);
  }
  
  return stream;
}

/* build the environment (OPT_env_files and OPT_env_chunks) */
static CtplEnviron *
build_environ (void)
{
  gboolean      success = TRUE;
  CtplEnviron  *env     = NULL;
  
  env = ctpl_environ_new ();
  /* load environment files */
  if (success && OPT_env_files) {
    gsize i;
    
    for (i = 0; success && OPT_env_files[i] != NULL; i++) {
      GError           *err = NULL;
      CtplInputStream  *stream;
      
      printv ("Loading environment file '%s'...\n", OPT_env_files[i]);
      stream = open_input_stream (OPT_env_files[i], &err);
      if (! stream ||
          ! ctpl_environ_add_from_stream (env, stream, &err)) {
        printerr ("Failed to load environment from file '%s': %s\n",
                  OPT_env_files[i], err->message);
        g_error_free (err);
        success = FALSE;
      }
      if (stream) {
        ctpl_input_stream_unref (stream);
      }
    }
  }
  /* load environment chunks */
  if (success && OPT_env_chunks) {
    gsize i;
    
    for (i = 0; success && OPT_env_chunks[i] != NULL; i++) {
      gchar  *chunk;
      GError *err = NULL;
      
      if (! encoding_needs_conversion (OPT_encoding)) {
        /* convert chunk to the requested encoding from utf8 (GLib converted the
         * argumlent to utf8) */
        chunk = g_convert (OPT_env_chunks[i], -1, OPT_encoding, "utf8",
                           NULL, NULL, &err);
      } else {
        /* no conversion needed, it's already in utf8 */
        chunk = g_strdup (OPT_env_chunks[i]);
      }
      printv ("Loading environment chunk '%s'...\n", chunk);
      if (! chunk || ! ctpl_environ_add_from_string (env, chunk, &err)) {
        printerr ("Failed to load environment from chunk '%s': %s\n",
                  chunk, err->message);
        g_error_free (err);
        success = FALSE;
      }
      g_free (chunk);
    }
  }
  if (! success) {
    ctpl_environ_unref (env);
    env = NULL;
  }
  
  return env;
}

/* parses a template from a file */
static gboolean
parse_template (const gchar      *filename,
                CtplOutputStream *output,
                CtplEnviron      *env,
                GError          **error)
{
  gboolean          rv = FALSE;
  CtplInputStream  *stream;
  
  stream = open_input_stream (filename, error);
  if (stream) {
    CtplToken *tree;
    
    tree = ctpl_lexer_lex (stream, error);
    ctpl_input_stream_unref (stream);
    if (tree) {
      rv = ctpl_parser_parse (tree, env, output, error);
    }
    ctpl_token_free (tree);
  }
  
  return rv;
}

/* parses all templates from OPT_input_files */
static gboolean
parse_templates (CtplEnviron       *env,
                 CtplOutputStream  *output)
{
  gboolean success = TRUE;
  
  if (OPT_input_files) {
    gsize i;
    
    for (i = 0; success && OPT_input_files[i] != NULL; i++) {
      GError *err = NULL;
      
      printv ("Parsing template '%s'...\n", OPT_input_files[i]);
      if (! parse_template (OPT_input_files[i], output, env, &err)) {
        printerr ("Failed to parse template '%s': %s\n",
                  OPT_input_files[i], err->message);
        g_error_free (err);
        success = FALSE;
      }
    }
  }
  
  return success;
}

static CtplOutputStream *
get_output_stream (void)
{
  GOutputStream    *gostream = NULL;
  CtplOutputStream *stream = NULL;
  
  if (OPT_output_file) {
    GFile              *file;
    GError             *err = NULL;
    GFileOutputStream  *gfostream;
    
    file = g_file_new_for_commandline_arg (OPT_output_file);
    gfostream = g_file_replace (file, NULL, FALSE, 0, NULL, &err);
    if (! gfostream) {
      printerr ("Failed to open output: %s\n", err->message);
      g_error_free (err);
    } else {
      gostream = G_OUTPUT_STREAM (gfostream);
    }
  } else {
    /* FIXME: how to get rid of GIOUnix for that? */
    gostream = g_unix_output_stream_new (STDOUT_FILENO, FALSE);
  }
  if (gostream) {
    if (encoding_needs_conversion (OPT_encoding)) {
      GCharsetConverter *converter;
      GError            *err = NULL;
      
      converter = g_charset_converter_new (OPT_encoding, "utf8", &err);
      if (! converter) {
        printerr ("Failed to create encoding converter: %s\n", err->message);
        g_error_free (err);
      } else {
        GOutputStream *gcostream;
        
        gcostream = g_converter_output_stream_new (gostream,
                                                   G_CONVERTER (converter));
        g_object_unref (gostream);
        gostream = gcostream;
        g_object_unref (converter);
      }
    }
    stream = ctpl_output_stream_new (gostream);
    g_object_unref (gostream);
  }
  
  return stream;
}


int main (int     argc,
          char  **argv)
{
  int     err   = 1;
  GError *error = NULL;
  
  setlocale (LC_ALL, "");
  g_type_init ();
  
  if (! parse_options (&argc, &argv, &error)) {
    printerr ("Option parsing failed: %s\n", error->message);
    g_clear_error (&error);
    err = 1;
  } else {
    CtplEnviron  *env;
    
    env = build_environ ();
    if (! env) {
      err = 1;
    } else {
      CtplOutputStream *ostream = get_output_stream ();
      
      if (ostream) {
        if (parse_templates (env, ostream)) {
          err = 0;
        }
        ctpl_output_stream_unref (ostream);
      }
      ctpl_environ_unref (env);
    }
  }
  
  return err;
}

